drivers/lora/sx12xx_common.c

```diff
@@ -144,6 +163,33 @@ int sx12xx_lora_recv(const struct device *dev, uint8_t *data, uint8_t size,
 	return dev_data.rx_len;
 }
 
+int sx12xx_lora_recv_async(const struct device *dev, lora_recv_cb cb)
+{
+	/* Handle RX cancel */
+	if (!cb) {
+		if (dev_data.rx_cb) {
+			/* Put radio to sleep */
+			Radio.Sleep();
+			dev_data.rx_cb = NULL;
+		}
+		return 0;
+	}
+
+	/* Handle reception already running */
+	if (dev_data.rx_cb) {
```

Comment:
This is going to be problematic :( What if both the ISR and the application tries to modify the callback? We are in a race, right? And do you really have a use case to switch the callbacks in runtime? Otherwise, please get rid of it.

Reply:
I don't have a use case for it, but thought someone might, some sort of packet state machine perhaps.

Yes this is technically a race condition, but so is every other function in this driver, none of them will work without external synchronization across multiple threads.

The idea was that once started, either the initiating thread OR the callback is responsible for future calls to `lora_recv_async`, but not both of them arbitrarily. As this is basically the same requirement as "2 threads cannot call `lora_recv` at the same time", I assumed it wouldn't be a problem.

Reply:
> Yes this is technically a race condition, but so is every other function in this driver, none of them will work without external synchronization across multiple threads.

Right, the reason was the use cases were simple at that point. But I'm pretty sure that we need to rewrite the driver in the future to handle concurrency when developers report issues or someone has time to look into it. But here, since the callback assignment is shared between the ISR and userspace, the issue is obvious. So I think you can just remove the callback reassigning part for now.